File: app/src/main/java/me/yxp/qfun/hook/api/OnTroopShutUp.java
package me.yxp.qfun.hook.api;

import org.json.JSONObject;

import java.lang.reflect.Method;

import me.yxp.qfun.hook.base.ApiHookItem;
import me.yxp.qfun.utils.hook.HookUtils;
import me.yxp.qfun.utils.json.ProtoData;
import me.yxp.qfun.utils.qq.QQUtils;
import me.yxp.qfun.utils.reflect.ClassUtils;
import me.yxp.qfun.utils.reflect.FieldUtils;
import me.yxp.qfun.utils.reflect.MethodUtils;

public final class OnTroopShutUp extends ApiHookItem {
    public static final ApiHookItem INSTANCE = new OnTroopShutUp();
    private static final String SERVICE_CMD = "trpc.msg.olpush.OlPushService.MsgPush";

    @Override
    public void loadHook() throws Throwable {
        Method onReceive = MethodUtils.create(ClassUtils._MSFServlet())
                .withMethodName("onReceive")
                .withParamTypes(ClassUtils._FromServiceMsg())
                .findOne();

        HookUtils.hookAlways(onReceive, null, param -> {
            Object fromServiceMsg = param.args[0];
            String serviceCmd = FieldUtils.create(fromServiceMsg)
                    .withName("serviceCmd")
                    .getValue()
                    .toString();

            if (!SERVICE_CMD.equals(serviceCmd)) {
                return;
            }

            byte[] wupBuffer = (byte[]) FieldUtils.create(fromServiceMsg)
                    .withName("wupBuffer")
                    .getValue();
            ProtoData data = new ProtoData();
            data.fromBytes(wupBuffer);
            JSONObject json = data.toJSON();
            JSONObject json1 = json.getJSONObject("1").getJSONObject("2");

            if (!(json1.getInt("1") == 732 && json1.getInt("2") == 12)) {
                return;
            }

            JSONObject json2 = json.getJSONObject("1").getJSONObject("3").getJSONObject("2");
            String troopUin = String.valueOf(json2.getLong("1"));
            String opUin = QQUtils.getUinFromUid(json2.getString("4"));
            String memberUin = QQUtils.getUinFromUid(json2.getJSONObject("5").getJSONObject("3").getString("1"));
            long time = json2.getJSONObject("5").getJSONObject("3").getLong("2");

            for (Listener listener : mListenerSet) {
                if (listener instanceof TroopShutUpListener) {
                    ((TroopShutUpListener) listener).onShutUp(troopUin, memberUin, time, opUin);
                }
            }
        });
    }

    public interface TroopShutUpListener extends Listener {
        void onShutUp(String troopUin, String memberUin, long time, String opUin) throws Throwable;
    }
}

File: app/src/main/java/me/yxp/qfun/hook/api/OnTroopShutUp.kt
package me.yxp.qfun.hook.api

import com.tencent.qphone.base.remote.FromServiceMsg
import me.yxp.qfun.annotation.HookItemAnnotation
import me.yxp.qfun.hook.base.BaseApiHookItem
import me.yxp.qfun.hook.base.Listener
import me.yxp.qfun.utils.hook.hookAfter
import me.yxp.qfun.utils.json.ProtoData
import me.yxp.qfun.utils.json.long
import me.yxp.qfun.utils.json.num
import me.yxp.qfun.utils.json.obj
import me.yxp.qfun.utils.json.str
import me.yxp.qfun.utils.json.walk
import me.yxp.qfun.utils.qq.FriendTool
import mqq.app.MSFServlet

@HookItemAnnotation("监听群禁言")
object OnTroopShutUp : BaseApiHookItem<TroopShutUpListener>() {

    private const val SERVICE_CMD = "trpc.msg.olpush.OlPushService.MsgPush"

    override fun loadHook() {

        MSFServlet::class.java
            .getDeclaredMethod("onReceive", FromServiceMsg::class.java)
            .hookAfter(this) { param ->

                val fromServiceMsg = param.args[0] as FromServiceMsg
                val cmd = fromServiceMsg.serviceCmd

                if (cmd != SERVICE_CMD) return@hookAfter

                val wupBuffer = fromServiceMsg.wupBuffer
                val data = ProtoData().apply { fromBytes(wupBuffer) }
                val json = data.toJSON()

                val msgHead = json.walk("1", "2")?.obj ?: return@hookAfter
                if (msgHead["1"].num != 732 || msgHead["2"].num != 12) {
                    return@hookAfter
                }

                val groupInfo = json.walk("1", "3", "2")?.obj ?: return@hookAfter

                val troopUin = groupInfo["1"].long?.toString() ?: return@hookAfter
                val opUid = groupInfo["4"].str ?: ""
                val opUin = FriendTool.getUinFromUid(opUid)

                val shutUpInfo = groupInfo.walk("5", "3")?.obj ?: return@hookAfter

                val memberUid = shutUpInfo["1"].str ?: ""
                val memberUin = FriendTool.getUinFromUid(memberUid)
                val time = shutUpInfo["2"].long ?: 0L

                listenerSet.filter { verify(it) }
                    .forEach {
                        it.onShutUp(troopUin, memberUin, time, opUin)
                    }
            }
    }
}

fun interface TroopShutUpListener : Listener {
    fun onShutUp(troopUin: String, memberUin: String, time: Long, opUin: String)
}